### DISCUSSION TITLE
Saleor for external authentication and data synch

### DISCUSSION THREAD

**Discussion by looped-monk** [saleor/saleor#10522] [General] [answered]
Hello :

My use case : I currently have a Django Blogging app as a backend with postid(product post id in case it is a shop blog post) and user details and I wish to enable my users to purchase products on my blog by connecting and pass data to Saleor for 
1) Authentication using JWT
2) Post id and other product data like product category etc and enable real time event synch so that data is available on both ends(Saleor & my Django Backend as well)

I intend to make Saleor handle the ecommerce journey of my users since its a fantastic platform to start with. 

I intend to also know :

3)My frontend blog app(flutter) needs to connect to my Django backend first and pass data to Saleor? or connect to Saleor and pass data to my Django backend? Which is the best approach for my use case considering the way Saleor is built with plugins and integrations ?

**Comment by karolkielecki** [comment]
Hi @narayanan-ka, thanks for your questions. Plugins will be deprecated soon. In the case of extensibility, we move towards apps. 
You can find more info in our documentation:
- https://docs.saleor.io/docs/3.x/developer/extending/apps/key-concepts
- https://docs.saleor.io/docs/3.x/dashboard/apps

Maybe that's not your use case, but you can also read more about how to combine headless architectures on our blog - Saleor x Strapi (open source CMS): https://saleor.io/blog/combining-headless-architectures-with-saleor-and-strapi-147/

**Reply by looped-monk** [reply]
Hello. I went through apps section. While its certainly a great feature, it certainly does not serve my use case. 

My understanding:


- The metadata api would make more sense since i need to update the Saleor with custom fields using mutations without touching the core and open up the possibility of bricking it. This would however be after saleor successfully authenticates my Django blogging app users.

- Furthur to above, I currently use JWT for my django app for password reset and email verification. It's best to extend this jwt authentication mechanism to pass data to Saleor for authentication. 

I choose  Saleor as a resource server method to handle this(since the login  logout, refresh tokens will be handled by my django app), but can't seem to figure how to use Open id on Saleor's end while my endpoint returns a jwt token containing all the payload to be passed to Saleor for successful authentication. Any help appreciated. 

@karolkielecki

**Reply by looped-monk** [reply]
@karolkielecki Will a JWT token would suffice or would I need to install and configure OIDC on my Django blogging apps end ?

**Comment by korycins** [comment]
Hey @narayanan-ka, If I understand correctly your case:

> Authentication using JWT

I would consider having an OIDC service, that would allow you to log in with the same JWT token to your App and Saleor instance. You can consider using the OIDC plugin for that: https://docs.saleor.io/docs/3.x/developer/available-plugins/openid-connect#saleor-as-a-resource-server (I know that Karol mentioned that plugins will be deprecated but I believe that we will first migrate OIDC plugin into core code.)

Another option is to treat Saleor as an authorization server. You can  do this with OIDC plugin also: https://docs.saleor.io/docs/3.x/developer/available-plugins/openid-connect#saleor-as-a-client-for-authorization-server

Also, Saleor's built-in JWT tokens are signed with RS256 (asymetric), https://docs.saleor.io/docs/3.x/developer/users#rs256-public-key, so your app could be able to validate Saleor's built-in JWT token and confirm that user is logged-in on Saleor side. https://auth0.com/blog/navigating-rs256-and-jwks/

> Post id and other product data like product category etc and enable real time event synch so that data is available on both ends(Saleor & my Django Backend as well)

According to the sync between apps. From the Saleor side, you can create an app instance, that will have its own authentication token: https://docs.saleor.io/docs/3.x/dashboard/apps. And then subscribe to the webhooks like product_updated/ category_updated etc: https://docs.saleor.io/docs/3.x/developer/extending/apps/asynchronous-webhooks. This will send any update events that happen on Saleor side to your app (you need to have an implementation for handling this on your side).
For the communication: Blog App-> Saleor, you can use a token (recieved during the creation of the app), to call mutations that will sync objects. 

> My frontend blog app(flutter) needs to connect to my Django backend first and pass data to Saleor? or connect to Saleor and pass data to my Django backend? Which is the best approach for my use case considering the way Saleor is built with plugins and integrations ?

I think that it depends on the implementation of the authorization layer. Like your app can be an authorization instance also, or Saleor can be. Or you can use an external OIDC provider like Auth0.

**Reply by looped-monk** [reply]
@korycins Thank you for taking time to answer this. It was detailed and well structured. I have one last question

How do I receive all notifications from Saleor side on My Django Blogging app end for product updates ?

**Reply by korycins** [reply, marked as answer]
You [create an App](https://docs.saleor.io/docs/3.x/dashboard/apps#creating-new-app), add a new webhook with an event like `PRODUCT_UPDATED`, then any changes in `Product` will trigger a webhook with payload (make sure that you add `manage_product` permission to your App). Same thing you can do for `ProductVariant`. 
Here you can find [all webhook events](https://docs.saleor.io/docs/3.x/developer/extending/apps/asynchronous-webhooks#available-webhook-events) (You can also find them in graphql's schema:  enums: `WebhookEventTypeSyncEnum` and `WebhookEventTypeAsyncEnum`)
You can change the shape of the payload that Saleor will send to your App: https://docs.saleor.io/docs/3.x/developer/extending/apps/subscription-webhook-payloads#custom-payloads

**Reply by looped-monk** [reply]
Thanks again.

**Reply by looped-monk** [reply]
@korycins Need a little bit of your help here. I couldn't figure out this one statement as part of Resource server configuration. 

""To use this flow you need to provide values for JSON Web Key Set URL and User info URL on the Saleor Dashboard side""

1)Since I will be Receiving the JWT token from my Django blogging backend app and passing it on to the Saloer backend from my front end flutter app, the Saleor backend will be validating the same with my Django blog Backend endpoints from where I got the necessary details  ?  

2)Also , I setup saleor docker and couldn't find the 2 places mentioned above to enter my backend details.

3)Can the above process be done by setting up an App on Saleor end ? - just thinking.

**Reply by korycins** [reply]
> 1)Since I will be Receiving the JWT token from my Django blogging backend app and passing it on to the Saloer backend from my front end flutter app, the Saleor backend will be validating the same with my Django blog Backend endpoints from where I got the necessary details ?

Your OIDC provider needs to support asymmetric signature keys. The public part should be available as JWKS: https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets. These keys will be used to validate token that we receive from your app.  The user info is the url used to fetch user details by using provided access token. It allows us to match recieved access token with the user's email. https://connect2id.com/products/server/docs/api/userinfo

2&3
OIDC plugin is configurable from dashboard side. Go to configuration->plugins-> OIDC plugin.
